lib/github.ml

```diff
@@ -304,15 +304,65 @@ let check_tag ~dry_run vcs tag =
   if Vcs.tag_exists ~dry_run vcs tag then Ok ()
   else
     R.error_msgf
-      "CHANGES.md lists '%s' as the latest release, but no corresponding tag \
+      "CHANGES.md lists '%a' as the latest release, but no corresponding tag \
        has been found in the repository.@.Did you forget to call 'dune-release \
        tag' ?"
-      tag
+      Vcs.Tag.pp tag
 
 let assert_tag_exists ~dry_run tag =
   Vcs.get () >>= fun repo ->
   if Vcs.tag_exists ~dry_run repo tag then Ok ()
-  else R.error_msgf "%s is not a valid tag" tag
+  else R.error_msgf "%a is not a valid tag" Vcs.Tag.pp tag
+
+let is_annotated_tag vcs remote_rev =
+  (* Resolve again in case of annotated tags (most common case).
+     This is a no-op for non-annotated tags. In case of error, we
+     can assume that the remote is different because we checked that we
+     have the tag locally. *)
+  match Vcs.commit_id ~commit_ish:remote_rev vcs with
+  | Ok local_rev ->
+      if local_rev = remote_rev then
+        (* if resolving again yields the same tag, then the tag is not annotated *)
+        false
+      else true
+  | Error _ -> false
+
+let validate_remote_tag vcs local_rev tag remote_rev_unpeeled =
+  let missing_tag pp_r =
+    App_log.unhappy (fun l ->
+        l
+          "The tag %a is present on the remote but points to a different \
+           commit (%a)."
+          Text.Pp.tag tag pp_r ())
+  in
+  match Vcs.commit_id ~commit_ish:remote_rev_unpeeled vcs with
+  | Ok remote_rev
+    when remote_rev = local_rev && is_annotated_tag vcs remote_rev_unpeeled ->
```

Comment:
The reason why it is calling `Vcs.commit_id` is because for extracting a function like `is_annotated_tag` I think it makes sense that the function would get the "unpeeled" commit, resolve it, and see itself. Yet here it calls `Vcs.commit_id` to see if there is a commit to begin with. While it is a bit ugly that it ends up calling `Vcs.commit_id` effectively twice I think it makes the code more understandable than before since the intent is now clearer.

The alternative is basically rewriting the function to

```ocaml
let is_annotated_tag tag resolved_tag = String.equal tag annotated_tag
```
(excuse the ETA-expansion for clarity)

which would work but the function would be somewhat nonsensical because it doesn't determine anything but rather requires that the candidate annotated and the resolved tag are passed to it.

Reply:
Ok I get it now!

I think a good alternative is to extract a function that will get the `remote_rev` and the `local_rev`, do the local de-referencing and return a value that carries all the information, something like ``[`Up_to_date_annotated | `Up_to_date_non_annotated | `Points_to_different_commit | `Points_to_missing_object]``.

Reply:
That's a great idea. Also rerenamed the functions to be closer to the original name and less vague.